File: test/unit/vbo.test.js
import { describe, expect, it } from "vitest";
import { parseVboText } from "../../public/js/import/vbo.js";
import { buildGate, deriveLaps, projectTrace } from "../../public/js/import/geo.js";
import { LAP_S, buildVboText, circleTrace } from "../fixtures/build.mjs";

const LAP_MS = Math.round(LAP_S() * 1000);

describe("parseVboText", () => {
  it("parses the created date, GPS points and duration", () => {
    const points = circleTrace();
    const out = parseVboText(buildVboText(points));
    expect(out.kind).toBe("vbo");
    expect(out.date).toBe("2026-06-20");
    expect(out.time).toBe("09:15:00");
    expect(out.gps.length).toBe(points.length);
    expect(out.durationS).toBeCloseTo(points[points.length - 1].t, 1);
    // minutes -> degrees, and Racelogic's west-positive longitude comes back
    // east-positive like every other source, so the racing line isn't mirrored
    expect(out.gps[0].lat).toBeCloseTo(points[0].lat, 4);
    expect(out.gps[0].lon).toBeCloseTo(points[0].lon, 4);
  });

  it("normalizes the km/h velocity column to m/s", () => {
    // circleTrace speed is 40 m/s; the fixture writes it as 144 km/h with a
    // "velocity kmh" header line. gps.v must come back in m/s — the per-lap
    // channel data (channels.js) converts m/s -> km/h and would otherwise
    // store speeds 3.6x too high.
    const out = parseVboText(buildVboText(circleTrace()));
    expect(out.gps[10].v).toBeCloseTo(40, 2);
  });

  it("computes laps from a [laptiming] start line", () => {
    const out = parseVboText(buildVboText(circleTrace(), { withLapTiming: true }));
    expect(out.needsLine).toBe(false);
    expect(out.laps).toHaveLength(3);
    for (const lap of out.laps) {
      expect(Math.abs(lap.timeMs - LAP_MS)).toBeLessThan(200);
      expect(lap.estimated).toBe(true);
    }
  });

  it("reads a latitude-first [laptiming] line too", () => {
    const out = parseVboText(buildVboText(circleTrace(), { withLapTiming: true, latFirst: true }));
    expect(out.laps).toHaveLength(3);
  });

  it("times laps driven just past the end of a short start line", () => {
    // Track Precision's line is ~15 m and sits to one side; a lap driven a
    // few metres wide of its end must still be timed. A 14 m line spans
    // radius 293-307 m; drive the circle at 310 m.
    const LAP = (2 * Math.PI * 310) / 40;
    const out = parseVboText(buildVboText(circleTrace({ radius: 310 }), { withLapTiming: true, lineHalfM: 7 }));
    expect(out.laps).toHaveLength(3);
    for (const lap of out.laps) expect(Math.abs(lap.timeMs - LAP * 1000)).toBeLessThan(200);
  });

  it("takes the gate direction from the trace, so a clockwise session times too", () => {
    // The same circle driven clockwise crosses the line the other way.
    const cw = circleTrace().map((p, i, all) => ({ ...all[all.length - 1 - i], t: p.t }));
    const out = parseVboText(buildVboText(cw, { withLapTiming: true }));
    expect(out.laps).toHaveLength(3);
  });

  it("times the first and last laps of a recording started and stopped at the line", () => {
    // Track Precision starts recording just past the line and stops just
    // short of it, so neither end is ever seen crossed. The fixture's line is
    // at a quarter turn; keep three laps starting 0.3 s (12 m) after it.
    const q = LAP_S() / 4;
    const kept = circleTrace({ revolutions: 4 }).filter((p) => p.t > q + 0.3 && p.t < q + 3 * LAP_S() - 0.3);
    const t0 = kept[0].t;
    const out = parseVboText(buildVboText(kept.map((p) => ({ ...p, t: p.t - t0 })), { withLapTiming: true }));
    expect(out.laps).toHaveLength(3);
    for (const lap of out.laps) expect(Math.abs(lap.timeMs - LAP_MS)).toBeLessThan(200);
  });

  it("parses the Porsche Track Precision layout and its car channels", () => {
    const points = circleTrace();
    const out = parseVboText(
      buildVboText(points, { withLapTiming: true, trackPrecision: true }),
      "recording-2026-06-06-09-53-45.vbo"
    );
    // one column name per line, names with spaces
    expect(out.laps).toHaveLength(3);
    // the date is the file name's — "created at" is the export time
    expect(out.date).toBe("2026-06-06");
    expect(out.time).toBe("09:15:00");
    const ch = out.carChannels;
    expect(Object.keys(ch).sort()).toEqual(["brake", "gear", "latG", "rpm", "steering", "throttle"]);
    // the pedal fraction becomes a percentage
    expect(Math.max(...ch.throttle.map((p) => p.v))).toBeCloseTo(100, 0);
    // brake pressure becomes a percentage of the file's peak
    expect(Math.max(...ch.brake.map((p) => p.v))).toBeCloseTo(100, 5);
    expect(Math.min(...ch.brake.map((p) => p.v))).toBe(0);
    // true G (LatAcc_PTPA) over the /9.81 `latacc` column, as a magnitude
    expect(Math.max(...ch.latG.map((p) => p.v))).toBeCloseTo(0.9, 2);
    expect(new Set(ch.gear.map((p) => p.v))).toEqual(new Set([3, 4]));
    // bar -> kPa; the 3276.8 "no reading" sentinel is dropped
    expect(out.lapScalarChannels.tyreKpaLF[0].v).toBeCloseTo(210, 5);
    expect(out.lapScalarChannels.tyreKpaRF).toBeUndefined();
  });

  it("falls back to the export date when the name has none", () => {
    const out = parseVboText(buildVboText(circleTrace(), { trackPrecision: true }), "session.vbo");
    expect(out.date).toBe("2026-09-22");
  });

  it("asks for a line when there is no [laptiming] section", () => {
    const out = parseVboText(buildVboText(circleTrace()));
    expect(out.needsLine).toBe(true);
    expect(out.laps).toEqual([]);
    // ...and the user-picked line then yields the laps
    const trace = projectTrace(out.gps, out.gps[0]);
    const gate = buildGate(trace, Math.round(0.25 * LAP_S() * 10));
    expect(deriveLaps(trace, gate)).toHaveLength(3);
  });

  it("rejects files without the expected structure", () => {
    expect(() => parseVboText("not a vbo")).toThrow(/column names/);
    expect(() =>
      parseVboText("[column names]\nsats time lat long\n[data]\n008 091500.00 2193.6 4752.0")
    ).toThrow(/no usable GPS/);
  });
});

File: apps/android/core/src/test/kotlin/app/trackevolution/core/telemetry/VBOTest.kt
package app.trackevolution.core.telemetry

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

/**
 * `test/unit/vbo.test.js`'s cases that the contract doesn't already pin, plus
 * the Kotlin-only seams (the name dispatch, `Number()`, the review strings).
 *
 * The JS builds its inputs with `buildVboText`; this suite starts from the
 * committed contract files instead and edits them, so it needs no second copy
 * of the fixture builder.
 */
class VBOTest {

    private val noLine = VBOFixtures.text("vbox-noline.vbo")
    private val withLine = VBOFixtures.text("vbox-laptiming.vbo")
    private val trackPrecision = VBOFixtures.text("trackprecision-2026-06-06-09-53-45.vbo")

    @Test
    fun `parses the created date, GPS points and duration`() {
        val out = VBO.parseVboText(noLine)
        assertEquals(ParsedTelemetry.Kind.VBO, out.kind)
        assertEquals("2026-06-20", out.date)
        assertEquals("09:15:00", out.time)
        assertEquals(dataRows(noLine).size, out.gps!!.size)
        // Racelogic's west-positive longitude comes back east-positive.
        assertTrue(out.gps!![0].lon < 0, "longitude is negated into east-positive degrees")
    }

    @Test
    fun `normalizes the km-h velocity column to m-s`() {
        // The fixture writes 40 m/s as 144 km/h under a "velocity kmh" header.
        val out = VBO.parseVboText(noLine)
        assertEquals(40.0, out.gps!![10].v!!, 1e-9)
    }

    @Test
    fun `reads mph and knots velocity headers`() {
        val mph = VBO.parseVboText(noLine.replace("velocity kmh", "velocity mph"))
        assertEquals(144 * 0.44704, mph.gps!![10].v!!, 1e-9)
        val kts = VBO.parseVboText(noLine.replace("velocity kmh", "velocity knots"))
        assertEquals(144 * 0.514444, kts.gps!![10].v!!, 1e-9)
    }

    @Test
    fun `takes the gate direction from the trace, so a clockwise session times too`() {
        // The same circle driven the other way: coordinates reversed, clock kept.
        val rows = dataRows(withLine).map { it.split(Regex("\\s+")) }
        val reversed = rows.indices.map { i ->
            val coords = rows[rows.size - 1 - i]
            listOf(rows[i][0], rows[i][1], coords[2], coords[3], coords[4]).joinToString(" ")
        }
        val head = withLine.substringBefore("[data]")
        val out = VBO.parseVboText(head + "[data]\n" + reversed.joinToString("\n") + "\n")
        assertEquals(3, out.laps.size)
        assertFalse(out.needsLine)
    }

    @Test
    fun `falls back to the export date when the name has none`() {
        assertEquals("2026-09-22", VBO.parseVboText(trackPrecision, "session.vbo").date)
        // …and the name wins when it has one, the export line being no recording's date.
        assertEquals("2026-06-06", VBO.parseVboText(trackPrecision, "recording-2026-06-06-14-59-38.vbo").date)
        // No "created on" line: the time is the first sample's time of day.
        assertEquals("09:15:00", VBO.parseVboText(trackPrecision, "session.vbo").time)
    }

    @Test
    fun `parses the Porsche Track Precision layout and its car channels`() {
        val out = VBO.parseVboText(trackPrecision, "recording-2026-06-06-09-53-45.vbo")
        val ch = out.carChannels
        val present = TelemetryChannels.CHANNEL_NAMES.map { it.first }.filter { ch[it] != null }.sorted()
        assertEquals(listOf("brake", "gear", "latG", "rpm", "steering", "throttle"), present)
        // the pedal fraction becomes a percentage
        assertEquals(100.0, ch.throttle!!.maxOf { it.v }, 0.5)
        // brake pressure becomes a percentage of the file's peak
        assertEquals(100.0, ch.brake!!.maxOf { it.v }, 1e-5)
        assertEquals(0.0, ch.brake!!.minOf { it.v })
        // true G (LatAcc_PTPA) over the /9.81 `latacc` column, as a magnitude
        assertEquals(0.9, ch.latG!!.maxOf { it.v }, 0.005)
        assertEquals(setOf(3.0, 4.0), ch.gear!!.map { it.v }.toSet())
        // bar -> kPa; the 3276.8 "no reading" sentinel is dropped
        assertEquals(210.0, out.lapScalarChannels["tyreKpaLF"]!![0].v, 1e-5)
        assertNull(out.lapScalarChannels["tyreKpaRF"])
        // the all-zero yaw column is no channel at all
        assertNull(ch.yaw)
    }

    @Test
    fun `asks for a line when there is no laptiming section`() {
        val out = VBO.parseVboText(noLine)
        assertTrue(out.needsLine)
        assertEquals(emptyList<ParsedLap>(), out.laps)
    }

    @Test
    fun `rejects files without the expected structure`() {
        val noColumns = assertThrows<TelemetryParseException> { VBO.parseVboText("not a vbo") }
        assertTrue(noColumns.message.contains("column names"), noColumns.message)
        val noGps = assertThrows<TelemetryParseException> {
            VBO.parseVboText("[column names]\nsats time lat long\n[data]\n008 091500.00 2193.6 4752.0")
        }
        assertTrue(noGps.message.contains("no usable GPS"), noGps.message)
        val missing = assertThrows<TelemetryParseException> {
            VBO.parseVboText("[column names]\nsats time lat\n[data]\n008 091500.00 2193.6")
        }
        assertTrue(missing.message.contains("time/lat/long"), missing.message)
    }

    @Test
    fun `computes an elevation range from a height column`() {
        val rows = dataRows(noLine).mapIndexed { i, row -> "$row ${100 + (i % 40)}" }
        val head = noLine.substringBefore("[data]").replace("sats time lat long velocity", "sats time lat long velocity height")
        val out = VBO.parseVboText(head + "[data]\n" + rows.joinToString("\n"))
        assertEquals(39.0, out.sessionMeta!!.elevationM!!, 1e-9)
    }

    // ---- Kotlin-only seams ----------------------------------------------------

    @Test
    fun `dispatches by file name, not by content`() {
        val bytes = VBOFixtures.bytes("vbox-laptiming.vbo")
        val out = Telemetry.parseTelemetryFile(ByteArraySource(bytes), "SESSION.VBO")
        assertEquals(ParsedTelemetry.Kind.VBO, out.kind)
        // attachLapChannels ran: the laps carry channel arrays.
        assertEquals(3, out.lapChannels!!.laps.size)
        // Without the name it is treated as video and refused as one.
        assertThrows<TelemetryParseException> { Telemetry.parseTelemetryFile(ByteArraySource(bytes)) }
    }

    @Test
    fun `Number() semantics`() {
        assertEquals(3.5, VBO.number("+003.5"))
        assertEquals(-0.0, VBO.number("-0"))
        assertEquals(0.0, VBO.number(""))
        assertEquals(26.0, VBO.number("0x1A"))
        assertEquals(5.0, VBO.number("5."))
        assertEquals(0.5, VBO.number(".5"))
        assertEquals(1500.0, VBO.number("1.5e3"))
        assertEquals(Double.NEGATIVE_INFINITY, VBO.number("-Infinity"))
        for (bad in listOf("1f", "1d", "NaN", "0x1p3", "1,5", "-0x10", "abc", ".")) {
            assertTrue(VBO.number(bad).isNaN(), "Number(\"$bad\") is NaN")
        }
    }

    @Test
    fun `reads the review strings the web writes for a vbo`() {
        val out = VBOFixtures.parse("vbox-laptiming.vbo")
        assertEquals("VBO 09:15:00", Telemetry.defaultLabel(out, "vbox-laptiming.vbo"))
        assertEquals(
            "Imported from vbox-laptiming.vbo — lap times derived from GPS start/finish crossings (~±0.1–0.3s)",
            Telemetry.importNotes(out, "vbox-laptiming.vbo"),
        )
    }

    private fun dataRows(text: String): List<String> =
        text.substringAfter("[data]").lines().map { it.trim() }.filter { it.isNotEmpty() }
}
